functions/supplier/cdc-debezium-supplier2/src/main/java/org/springframework/cloud/fn/supplier/cdc/CdcConfiguration.java

```diff
@@ -0,0 +1,225 @@
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.beans.factory.BeanClassLoaderAware;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.function.context.FunctionProperties;
+import org.springframework.cloud.stream.function.StreamBridge;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.MimeTypeUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ *
+ * @author Christian Tzolov
+ */
+@Configuration
+@EnableConfigurationProperties(CdcProperties.class)
+public class CdcConfiguration implements BeanClassLoaderAware {
+
+	private static final Log logger = LogFactory.getLog(CdcConfiguration.class);
+
+	/**
+	 * ORG_SPRINGFRAMEWORK_KAFKA_SUPPORT_KAFKA_NULL.
+	 */
+	public static final String ORG_SPRINGFRAMEWORK_KAFKA_SUPPORT_KAFKA_NULL = "org.springframework.kafka.support.KafkaNull";
+
+	private Object kafkaNull = null;
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		try {
+			Class<?> clazz = ClassUtils.forName(ORG_SPRINGFRAMEWORK_KAFKA_SUPPORT_KAFKA_NULL, classLoader);
+			Field field = clazz.getDeclaredField("INSTANCE");
+			this.kafkaNull = field.get(null);
+		}
+		catch (ClassNotFoundException | NoSuchFieldException | IllegalAccessException e) {
+		}
+	}
+
+	@Bean
+	public Properties cdcConfiguration(CdcProperties properties) {
+		Properties outProps = new java.util.Properties();
+		outProps.putAll(properties.getDebezium());
+		return outProps;
+	}
+
+	@Bean
+	@ConditionalOnProperty(name = "cdc.format", havingValue = "json", matchIfMissing = true)
+	public DebeziumEngine<?> debeziumEngineJson(Consumer<ChangeEvent<String, String>> changeEventConsumer,
+			java.util.Properties cdcConfiguration) {
+
+		DebeziumEngine<ChangeEvent<String, String>> debeziumEngine = DebeziumEngine
+				.create(io.debezium.engine.format.Json.class)
+				.using(cdcConfiguration)
+				.notifying(changeEventConsumer)
+				.build();
+
+		return debeziumEngine;
+	}
+
+	@Bean
+	@ConditionalOnProperty(name = "cdc.format", havingValue = "avro")
+	public DebeziumEngine<?> debeziumEngineAvro(Consumer<ChangeEvent<byte[], byte[]>> changeEventConsumer,
+			java.util.Properties cdcConfiguration) {
+
+		DebeziumEngine<ChangeEvent<byte[], byte[]>> debeziumEngine = DebeziumEngine
+				.create(io.debezium.engine.format.Avro.class)
+				.using(cdcConfiguration)
+				.notifying(changeEventConsumer)
+				.build();
+
+		return debeziumEngine;
+	}
+
+	@Bean
+	public EmbeddedEngineExecutorService embeddedEngine(DebeziumEngine<?> debeziumEngine) {
+
+		return new EmbeddedEngineExecutorService(debeziumEngine) {
+			@PostConstruct
```

Comment:
This is not OK to perform some low-level resource connection in the initialization phase.
Looks like this `EmbeddedEngineExecutorService` is our class, so let's consider to implement a `SmartLifecycle` contract instead!

Reply:
you're right. This ugly. I will implement the smat lifecycle.

Reply:
resolved